File: internal/httpapi/handlers/activity.go
package handlers

import (
	"github.com/RoseKhlifa/FarmBot/internal/domain/activity"
	"github.com/gin-gonic/gin"
)

func (h *Handler) RegisterActivity(r gin.IRouter) {
	r.GET("/api/activity/list", h.activityList)
	r.GET("/api/activity/group/:id", h.activityGroup)
	r.GET("/api/server-version", h.serverVersion)
	r.GET("/api/activity/helu", h.helu)
	r.POST("/api/activity/helu/draw", h.heluDraw)
	r.POST("/api/activity/helu/passport/claim", h.heluPassport)
	r.POST("/api/activity/helu/solar/claim", h.heluSolar)
	r.POST("/api/activity/helu/exchange", h.heluExchange)
	r.POST("/api/activity/qingmei/claim", h.qingmeiClaim)
	r.POST("/api/activity/qingmei/wine/sell", h.qingmeiWine)
	r.GET("/api/activity/guanxing", h.guanxing)
	r.POST("/api/activity/guanxing/claim", h.guanxingClaim)
	r.GET("/api/activity/shop", h.activityShop)
	r.POST("/api/activity/shop/buy", h.activityShopBuy)
	r.POST("/api/activity/shop/refresh", h.activityShopRefresh)
}
func (h *Handler) activityService(c *gin.Context) (activity.Service, bool) {
	id, ok := accountID(c, true)
	if !ok {
		return nil, false
	}
	service, ok := resolve(c, h.app().Domains.Activity, id)
	return service, ok
}
func (h *Handler) activityList(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	data, err := service.ListActivities(c.Request.Context())
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) activityGroup(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	id, ok := int64Param(c, "id")
	if !ok {
		return
	}
	data, err := service.GetActivityGroup(c.Request.Context(), id, c.Query("uid"))
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) serverVersion(c *gin.Context) { writeOK(c, gin.H{"version": "go"}) }
func (h *Handler) helu(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	data, err := service.GetHeluActivity(c.Request.Context())
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) heluDraw(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	data, err := service.DrawHeluGiftLotus(c.Request.Context(), activity.HeluDrawOptions{})
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) heluPassport(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	data, err := service.ClaimSeasonPassportRewards(c.Request.Context())
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) heluSolar(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	var body struct {
		ID int64 `json:"id"`
	}
	if !bindJSON(c, &body) {
		return
	}
	data, err := service.ClaimSolarTermsReward(c.Request.Context(), body.ID)
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) heluExchange(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	var body struct{ GoodsID, Count int64 }
	if !bindJSON(c, &body) {
		return
	}
	data, err := service.ExchangeHeluShopItem(c.Request.Context(), body.GoodsID, body.Count)
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) qingmeiClaim(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	data, err := service.ClaimQingmeiSeeds(c.Request.Context())
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) qingmeiWine(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	data, err := service.BrewAndSellQingmeiWine(c.Request.Context(), activity.QingmeiWineOptions{})
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) guanxing(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	data, err := service.GetGuanxingActivity(c.Request.Context())
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) guanxingClaim(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	data, err := service.ClaimGuanxingRewards(c.Request.Context())
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) activityShop(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	data, err := service.GetNanguaShop(c.Request.Context())
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) activityShopBuy(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	var body struct{ SlotID, Count int64 }
	if !bindJSON(c, &body) {
		return
	}
	data, err := service.BuyNanguaShopItem(c.Request.Context(), body.SlotID, body.Count)
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}
func (h *Handler) activityShopRefresh(c *gin.Context) {
	service, ok := h.activityService(c)
	if !ok {
		return
	}
	data, err := service.RefreshNanguaShop(c.Request.Context())
	if err != nil {
		writeError(c, err)
		return
	}
	writeOK(c, data)
}

File: web/src/api/activity.ts
import type { ApiBody, ApiEnvelope, ApiRequestConfig, Identifier } from './types'
import client from './client'
import { pathSegment } from './types'

export function listActivities<T = any>(config?: ApiRequestConfig) {
  return client.get<ApiEnvelope<T>>('/api/activity/list', config)
}

export function getActivityGroup<T = any>(id: Identifier, uid?: Identifier, config?: ApiRequestConfig) {
  return client.get<ApiEnvelope<T>>(`/api/activity/group/${pathSegment(id)}`, {
    ...config,
    params: { ...config?.params, ...(uid === undefined ? {} : { uid }) },
  })
}

export function getServerVersion<T = any>(config?: ApiRequestConfig) {
  return client.get<ApiEnvelope<T>>('/api/server-version', config)
}

export function getHelu<T = any>(config?: ApiRequestConfig) {
  return client.get<ApiEnvelope<T>>('/api/activity/helu', config)
}

export function drawHelu<T = any>(payload: ApiBody = {}, config?: ApiRequestConfig) {
  return client.post<ApiEnvelope<T>>('/api/activity/helu/draw', payload, config)
}

export function claimHeluPassport<T = any>(config?: ApiRequestConfig) {
  return client.post<ApiEnvelope<T>>('/api/activity/helu/passport/claim', {}, config)
}

export function claimHeluSolar<T = any>(payload: { id: number }, config?: ApiRequestConfig) {
  return client.post<ApiEnvelope<T>>('/api/activity/helu/solar/claim', payload, config)
}

export function exchangeHelu<T = any>(payload: { goodsId: number, count: number }, config?: ApiRequestConfig) {
  return client.post<ApiEnvelope<T>>('/api/activity/helu/exchange', payload, config)
}

export function claimQingmei<T = any>(config?: ApiRequestConfig) {
  return client.post<ApiEnvelope<T>>('/api/activity/qingmei/claim', {}, config)
}

export function sellQingmeiWine<T = any>(payload: ApiBody = {}, config?: ApiRequestConfig) {
  return client.post<ApiEnvelope<T>>('/api/activity/qingmei/wine/sell', payload, config)
}

export function getGuanxing<T = any>(config?: ApiRequestConfig) {
  return client.get<ApiEnvelope<T>>('/api/activity/guanxing', config)
}

export function claimGuanxing<T = any>(config?: ApiRequestConfig) {
  return client.post<ApiEnvelope<T>>('/api/activity/guanxing/claim', {}, config)
}

export function getActivityShop<T = any>(config?: ApiRequestConfig) {
  return client.get<ApiEnvelope<T>>('/api/activity/shop', config)
}

export function buyActivityShop<T = any>(payload: { slotId: number, count: number }, config?: ApiRequestConfig) {
  return client.post<ApiEnvelope<T>>('/api/activity/shop/buy', payload, config)
}

export function refreshActivityShop<T = any>(config?: ApiRequestConfig) {
  return client.post<ApiEnvelope<T>>('/api/activity/shop/refresh', {}, config)
}

export const getHeluActivity = getHelu
export const claimHeluSolarReward = claimHeluSolar
export const claimQingmeiSeeds = claimQingmei
export const brewAndSellQingmeiWine = sellQingmeiWine
